readthedocs/proxito/README.rst

```diff
@@ -42,6 +42,80 @@ path prefix:
+
+urlpattern_subproject:
+   URL pattern to change the path up to the subproject alias (the ``/projects`` prefix).
+   To change the pattern of the subproject itself, use ``urlpattern``.
+
+A URL pattern is a regex with replacement fields,
+replacement fields are expanded to define the proper regex and capture groups (components).
+Valid replacement fields are: ``language``, ``version``, ``filename``, and ``subproject``.
+
+Where to define the custom URL patterns
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To change the URL pattern of a project,
+define the custom URL pattern in the ``urlpattern`` attribute of the project itself,
+or if it's a translation, in the main language project.
+
+And to change the URL pattern of the subproject prefix (``/projects``),
+define the custom URL pattern in the  ``urlpattern_subproject`` attribute of the parent project.
+
+Writing a custom URL pattern
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. note::
+
+   Custom URL patterns aren't meant to be written by users (yet),
+   a Read the Docs developer needs to do it for the user.
+
+The simplest way to write a custom URL pattern is to start from the default ones,
+and change them as needed.
+
+- Multi version project: ``^/{language}(/({version}(/{filename})?)?)?$``
+- Single version project: ``^/{filename}$``
+- Subproject: ``^/projects/{subproject}(/{filename})?$``
+
+A couple of rules to be followed to avoid unexpected behaviours are:
+
+- Always put the filename at the end of the pattern,
+  this is since a filename matches any string.
+- Write the regex in a way that a partial match is possible,
+  this is useful for raising the proper exception on partial matches
+  (like a 404 on an invalid language if the URL starts with one).
+- Make the slashes optional (``/``) at the end of each component when possible,
+  this is since a URL is still valid if it doesn't end with a slash
+  (proxito will redirecte to the path with the slash when appropriate).
+- Always end the regex with ``$``, so the whole path is matched,
+  not just a part of the path.
+- Starting the regex with ``^`` is optional, since we use ``re.match``.
+- All paths are guaranteed to start with ``/``, so always start the regex with ``/``.
+- Don't include named groups, character classes (``[]``),
+  or anything that can't be used to reconstruct the URL from the pattern
+  using only the version, language, filename, and subproject components.
+- Is recomeded to separate each component with ``/``,
+  since any other characters can be valid in components (like ``-`` in slugs).
+- Remember that this is a regex, so make sure to escape special characters like ``.``.
+- Make sure that the patterns from the parent project and the subproject prefix
+  don't overlap (``urlpattern`` and ``urlpattern_subproject``).
+  We first try to match the path against ``urlpattern`` and then against ``urlpattern_subproject``,
+  if they overlap, we won't be able to serve subprojects.
+
```

Comment:
I'm definitely worried about how complex this is. I do think we need some kind of basic test suite or something that runs when we try to set these, otherwise I'm 99% sure we will break things. 

I understand why some of this complexity is needed, but I'd really like to find a way to make it less likely to break.